lib/please_cli/please_cli/decision_task.py

```diff
@@ -323,6 +344,197 @@ def get_task(task_group_id,
     }
 
 
+NixPath = str
+NixHash = str
+NixHashes = typing.List[typing.Tuple[NixPath, NixHash]]
+# TODO: we need to have more detailed type for Projects but we would need to
+#       have a type for please_cli.config.PROJECTS_CONFIG before that
+Projects = typing.Dict
+
+
+async def read_stream(stream,
+                      log_output: typing.Optional[typing.Callable[[str], None]] = None,
+                      callback: typing.Optional[typing.Callable[[str], None]] = None,
+                      ) -> str:
+    output = []
+    while True:
+        line = await stream.readline()
+        if line:
+            line = line.decode().rstrip('\n')
+            if log_output:
+                log_output(line)
+            output.append(line)
+            if callback:
+                callback(line)
+        else:
+            break
+    return '\n'.join(output)
+
+
+async def run(_command: typing.Union[str, typing.List[str]],
+              stream: bool = False,
+              handle_stream_line: typing.Optional[typing.Callable[[str], None]] = None,
+              log_command: bool = True,
+              log_output: bool = True,
+              secrets: typing.List[str] = [],
+              **kwargs,
+              ) -> typing.Tuple[int, str, str]:
```

Comment:
I would add a big comment explaining what's going on, the parameters, especially `**kwargs` being fed to suprocess